### PR TITLE
[Inductor] make sure the type signature always matches

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90626
* #90121
* __->__ #90120

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Repro (without this patch):
```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 pytest -rfxs -v test/inductor/test_torchinductor.py  -k test_upsample_bilinear2d_b_cuda --capture=no
[...]
  File "torch/_dynamo/output_graph.py", line 551, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e) from e
torch._dynamo.exc.BackendCompilerFailed: compile_fx_wrapper raised LoweringException: TypeError: sym_float() missing 1 required positional argument: 'a'
  target: <function sym_float at 0x7f3b86f6e200>
  args[0]: 2.0*s4

While executing %sym_float : [#users=2] = call_function[target=torch.fx.experimental.symbolic_shapes.sym_float](args = (%mul,), kwargs = {})
```

Repro (with this patch); more fixes required:

```
  File "torch/fx/experimental/symbolic_shapes.py", line 107, in sym_float
    return float(a)
  File "/lib/python3.10/site-packages/sympy-1.11.1-py3.10.egg/sympy/core/expr.py", line 351, in __float__
    raise TypeError("Cannot convert expression to float")
TypeError: Cannot convert expression to float
```

Codegen (printed from `torch.fx.graph._python_code`):
```
def forward(self, arg0_1):
    sym_size = torch.ops.aten.sym_size(arg0_1, 2)
    mul = sym_size * 2.0
    sym_float = torch.fx.experimental.symbolic_shapes.sym_float(mul);  mul = None
    floor = math_floor(sym_float)
    arange = torch.ops.aten.arange.default(floor, dtype = torch.float32, device = device(type='cuda', index=0), pin_memory = False);  floor = None
    sym_size_1 = torch.ops.aten.sym_size(arg0_1, 3)
    mul_1 = sym_size_1 * 2.0
    sym_float_1 = torch.fx.experimental.symbolic_shapes.sym_float(mul_1);  mul_1 = None
    floor_1 = math_floor(sym_float_1)
    arange_1 = torch.ops.aten.arange.default(floor_1, dtype = torch.float32, device = device(type='cuda', index=0), pin_memory = False);  floor_1 = None
[...]
```

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -2347,8 +2347,13 @@ def copy_input(x):
...
     @classmethod
     def process_kernel(cls, kernel, *args, **kwargs):
-        binded_args = signature(kernel).bind(*args, **kwargs).arguments
-        args_flat, args_spec = pytree.tree_flatten(binded_args)
+        # Define a wrapper so that the type signature always contains args and
+        # kwargs, which get accessed in unflatten_args.
+        def generic_kernel(*args, **kwargs):
+            return kernel(*args, **kwargs)
+
+        bound_args = signature(generic_kernel).bind(*args, **kwargs).arguments
```
Comment (nkaretnikov) [inline torch/_inductor/ir.py:2355]: Note: should also call `generic_kernel` where `kernel` is used below, just to be safe.

```
example_output = kernel(*new_args, **new_kwargs)
```

**Comment by ezyang** [comment]
So, no test because this isn't enough fixes?

**Comment by ezyang** [comment]
I don't really understand what the old code was doing, nor do I understand the new code. Adding other reviewers.

**Comment by ngimel** [comment]
If we indeed want to handle `torch.fx.experimental.symbolic_shapes.sym_float` via fallback kernel in inductor, then these changes are fine, I'm not sure that's the case though, cc @Chillee

**Comment by nkaretnikov** [comment]
@ngimel Note: this is not `sym_float`-specific. It just came up while trying to make `upsample_bilinear` decomp work.

This will fail for any function that doesn't have `*args` and `**kwargs` in the signature, if it happens to be propagated here. Are there any other legit examples of this yet in decomps? I dunno.

For example:

```
In [1]: from torch._inductor.ir import ExternKernel
In [2]: from torch.fx.experimental.symbolic_shapes import sym_float
In [3]: import sympy

# original issue that came up in decomp
In [4]: expr = sympy.Number(2.) * sympy.Symbol('foo', number=True)
In [5]: ExternKernel.process_kernel(sym_float, expr)
TypeError: sym_float() missing 1 required positional argument: 'a'

# similar error (made up example)
In [6]: ExternKernel.process_kernel(lambda x: x + 1, 7)
TypeError: <lambda>() missing 1 required positional argument: 'x'
```

This is where the problem happens in `unflatten_args`:
```
return result.get("args", []), result.get("kwargs", {})
```

This will just return empty lists if these parameters are not there. So in this PR I wrap the original call into a func with `*args` and `**kwargs` to generate the right signature.

In any case, silently moving on regardless of whether args/kwargs are there in the first place seems wrong to me. This was changed pretty recently, I think the author just didn't account for this corner case.

**Comment by ngimel** [comment]
I understand the issue is not specific to `sym_float` and would happen to any function that has a signature other than `*args, **kwargs`. However, fallback kernels are targeted specifically at `torch.ops.aten.*` ops that have this signature, and whether we want to expand them to everything else that might get in the trace is a question I would like more input on.

**Pull request closed without merging**